### PR TITLE
fix(web): key tip constraint logic requires .bottom CSS 

### LINKED ISSUE #6783: bug(ios): key tips for top row cropped off if no predictive-text banner is active
**Describe the bug**

Image from iOS

See the keytip (or lack thereof, practically speaking) in the screenshot above.

This was a regression likely introduced by #6383; it was not tested against a keyboard without a predictive text banner, so far as I can tell, which is how it went uncaught there.

**Expected behavior**
The key tip should always be constrained within the boundaries of the OSK.

---

**Keyman for iPhone/iPad/Android:**
 - Keyman version 15.0-stable, 16.0-alpha
 - Target application: any

### PR DESCRIPTION
**Pull request by jahorton**
Fixes #6783.

Key tips should properly remain within the OSK's bounds again after this fix.

Screen Shot 2022-06-16 at 10 50 57 AM

As this is a top-row key preview, note that its base key is completely obscured by the preview.

This fixes a regression introduced in #6383.  (Gotta test the top row _without_ predictive-text!)

# User Testing

These tests are derived from those of #6383, but they're a bit more precise and targeted than before.

### Test environments

In-app groups:
- GROUP_IPHONE_13_APP - use the Keyman app with a (simulated or real) iPhone 13.
- GROUP_IPHONE_SE_2_APP - use the Keyman app with a (simulated or real) iPhone SE 2nd gen.
- GROUP_ANDROID_APP - use the Keyman app with a (simulated or real) Android device - your pick.

In-browser groups:
- GROUP_IPHONE_13_BROWSER - use Safari on a (simulated or real) iPhone 13.
- GROUP_IPHONE_SE_2_BROWSER - use Safari on a (simulated or real) iPhone SE 2nd gen.
- GROUP_ANDROID_BROWSER - use Chrome on a (simulated or real) Android device - your pick.
    - Use the "Test unminified Keymanweb" Web test page.

### Tests

- TEST_TAMIL_TOP_PREVIEW:  Using the Tamil99 keyboard (`ekwtamil99uni`), verify that top row keypresses do not result in a cropped key preview.
    - The result may not be super-pretty, but the key thing is that it's uncropped.
    - There may be a little "stub" at the bottom of the preview; that's okay.
- TEST_TAMIL_NORMAL_PREVIEW:  Using the Tamil99 keyboard (`ekwtamil99uni`), verify that keypresses in the second to last row display nicely.
    - If there is a noticeable gap between the key preview's bottom and the bottom of its corresponding key, **_FAIL_** this test.
    - If the bottom border appears a bit large, but there is no _gap_, do not fail this test.  (Yay for subpixel-precision effects...)

- TEST_CHEROKEE_TOP_PREVIEW:  Using the Cherokee Nation (SIL) keyboard (`sil_cherokee_nation`), verify that top row keypresses do not result in a cropped key preview.
- TEST_CHEROKEE_NORMAL_PREVIEW:  Using the Cherokee Nation (SIL) keyboard (`sil_cherokee_nation`), verify that keypresses in the second to last row display nicely.

## SUITE_2:  predictive text active

This is only written as a separate suite because the predictive-text test can't be done with the (same) in-browser Web test page.

### Test environments

- GROUP_IPHONE_13_APP - use the Keyman app with a (simulated or real) iPhone 13.
- GROUP_IPHONE_SE_2_APP - use the Keyman app with a (simulated or real) iPhone SE 2nd gen.
- GROUP_ANDROID_APP - use the Keyman app with a (simulated or real) Android device - your pick.

### Test

- TEST_EUROLATIN_TOP_PREVIEW:  Using the default SIL EuroLatin keyboard, with predictive text active, verify that top row keypresses result in a _normal_ key preview.
    - If there is a noticeable gap between the key preview's bottom and the bottom of its corresponding key, **_FAIL_** this test.
    - If the bottom border appears a bit large, but there is no _gap_, do not fail this test.  (Yay for subpixel-precision effects...)

### CODE REVIEW

File: web/source/osk/browser/keytip.ts [hunk 1]
```diff
@@ -78,7 +78,7 @@ namespace com.keyman.osk.browser {
         let canvasWidth = xWidth + Math.ceil(xWidth * 0.3) * 2;
         let canvasHeight = Math.ceil(2.3 * xHeight) + (ySubPixelPadding); //
 
-        kts.top = Math.floor(y - canvasHeight) + 'px';
+        kts.bottom = Math.floor(keyman.osk.computedHeight - y) + 'px';
```
Comment (jahorton) [inline web/source/osk/browser/keytip.ts:81]: Why does this fix the issue?

https://github.com/keymanapp/keyman/blob/14b14857728d55948fa4b24b3ab67ac8438687f9/web/source/osk/browser/keytip.ts#L116-L119

Line 118 was broken - as it turns out, `cs.bottom == 'auto'` in our currently-released version.  That doesn't convert nicely to a number... causing `bottomY == NaN`.  That introduces a _lovely_ set of issues, as you can imagine.

https://github.com/keymanapp/keyman/blob/14b14857728d55948fa4b24b3ab67ac8438687f9/web/source/osk/browser/keytip.ts#L128-L129

The problem is, we _need_ `bottomY` for our calculations that constrain the keytip within bounds.  So, the best answer:  flip from a "top"-based approach to a "bottom"-based one.

-----

After a bit of analysis, I determined `y` to be "the distance of the key tip's bottom from the _top_ of its parent element."  Thing is, for CSS `.bottom` use, we need that to be "... from the _bottom_ of its parent element."  Fortunately, a simple calc.
-  Before, for use of `.top`, it needed "the distance of the key tip's top...".

**Review by mcdurdin — APPROVED** [review]
LGTM

**Comment by bharanidharanj** [comment]
## SUITE_1: no predictive text

### GROUP_IPHONE_13_APP: - use the Keyman app with a (simulated or real) iPhone 13.

* **TEST_TAMIL_TOP_PREVIEW (PASSED):** Tested Keyman 16.0.13 alpha-6784 build in iOS 15.0 / iPhone 13 along with Tamil99 Keyboard (ekwtami199uni), and it seems that the key preview is uncropped. 

 


* **TEST_TAMIL_NORMAL_PREVIEW (FAILED):** There is a noticeable gap (in the bottom layer)  between the key preview's bottom and the bottom of its corresponding key. Please, see the attached Screenshot. 
 


* **TEST_CHEROKEE_TOP_PREVIEW (PASSED):** Tested Keyman 16.0.13 alpha-6784 build in iOS 15.0 / iPhone 13 along with Cherokee Nation (SIL) Keyboard (sil_cherokee_nation) and it seems that the key preview is uncropped.
 



* **TEST_CHEROKEE_NORMAL_PREVIEW (FAILED):** Keypresses in the Second and last last row would display the Key preview nicely, but there is a gap between the Key Preview botton and the bottom of its corresponding key. Seems to be an issue.

**Comment by jahorton** [comment]
... wherein we gain yet another case of cleanup-mode "oh, I can just remove this one extra line I threw during development" turning out to be a "bad idea."

SUITE_1
GROUP_IPHONE_13_APP
TEST_TAMIL_NORMAL_PREVIEW OPEN
TEST_CHEROKEE_NORMAL_PREVIEW OPEN

**Comment by bharanidharanj** [comment]
### GROUP_ANDROID_APP: - use the Keyman app with a (simulated or real) Android device - your pick.

* **TEST_TAMIL_TOP_PREVIEW (PASSED):** Tested Keyman 16.0.13 alpha-6784 build in API 29 / Android 10 Simulator along with Tamil99 Keyboard (ekwtami199uni), and it seems that the key preview is uncropped. 



* **TEST_TAMIL_NORMAL_PREVIEW (PASSED):** There is no gap between the Key preview's bottm and the bottom of its corresponding key. 


* **TEST_CHEROKEE_TOP_PREVIEW (PASSED):** Tested Keyman 16.0.13 alpha-6784 build in API 29 / Android 10 Simulator along with Cherokee Nation (SIL) Keyboard (sil_cherokee_nation) and it seems that the key preview is uncropped.



* **TEST_CHEROKEE_NORMAL_PREVIEW (PASSED):** There is not gap between the Key preview's bottom and the bottom of its corresponding key.

**Comment by bharanidharanj** [comment]
## SUITE_2: predictive text active

### GROUP_ANDROID_APP: - use the Keyman app with a (simulated or real) Android device - your pick.

* **TEST_EUROLATIN_TOP_PREVIEW (PASSED):** Tested in Keyman 16.0.13 alpha test-6784 in API 29 / Android 10 Simulator and it seems to be working as expected.

**Comment by bharanidharanj** [comment]
### GROUP_IPHONE_13_APP: - use the Keyman app with a (simulated or real) iPhone 13.

* **TEST_TAMIL_NORMAL_PREVIEW (PASSED):** Retested with the updated Keyman 16.0.13-alpha build in iOS 15.0 / iPhone 13 Simulator along with Tamil99 Keyboard and it seems to be working fine. 

 

* **TEST_CHEROKEE_NORMAL_PREVIEW (PASSED):** Retested with the updated Keyman 16.0.13-alpha build in iOS 15.0 / iPhone 13 Simulator along with Cherokee Keyboard and it seems to be working fine.

**Comment by bharanidharanj** [comment]
## SUITE_1: no predictive text

### GROUP_IPHONE_13_APP: - use Safari on a (simulated or real) iPhone 13.

* **TEST_TAMIL_TOP_PREVIEW (PASSED):** Tested this in iOS 15.0 / iPhone 13 Simulator in Test unminified Keymanweb" Web test page using Safari Browser and it is working as expected. 
* **TEST_TAMIL_NORMAL_PREVIEW (PASSED):** Tested this in iOS 15.0 / iPhone 13 Simulator in Test unminified Keymanweb" Web test page using Safari Browser and it is working as expected.
 


* **TEST_CHEROKEE_TOP_PREVIEW (PASSED):** Tested this in iOS 15.0 / iPhone 13 Simulator in Test unminified Keymanweb" Web test page using Safari Browser and it is working as expected.
* **TEST_CHEROKEE_NORMAL_PREVIEW (PASSED):** Tested this in iOS 15.0 / iPhone 13 Simulator in Test unminified Keymanweb" Web test page using Safari Browser and it is working as expected.

**Comment by bharanidharanj** [comment]
## SUITE_1: no predictive text

### GROUP_IPHONE_SE_2_APP: - use Safari on a (simulated or real) iPhone SE 2nd gen.

* **TEST_TAMIL_TOP_PREVIEW (PASSED):** notes
* **TEST_TAMIL_NORMAL_PREVIEW (PASSED):** notes
* **TEST_CHEROKEE_TOP_PREVIEW (PASSED):** notes
* **TEST_CHEROKEE_NORMAL_PREVIEW (PASSED):** notes

**Comment by bharanidharanj** [comment]
## SUITE_2: predictive text active

### GROUP_IPHONE_13_APP: - use the Keyman app with a (simulated or real) iPhone 13.

* **TEST_EUROLATIN_TOP_PREVIEW (PASSED):** Tested in Keyman 16.0.13 alpha build in iOS 15.0 / iPhone 13 Simulator and it is working as expected. 

### GROUP_IPHONE_SE_2_APP: - use the Keyman app with a (simulated or real) iPhone SE 2nd gen.

* **TEST_EUROLATIN_TOP_PREVIEW (PASSED):** Tested in Keyman 16.0.13 alpha build in iOS 15.0 / SE 2nd gen Simulator and it is working as expected.

**Comment by bharanidharanj** [comment]
## SUITE_1: no predictive text

### GROUP_ANDROID_APP: - use Chrome on a (simulated or real) Android device - your pick.

* **TEST_TAMIL_TOP_PREVIEW (FAILED):** Tested this in API 29 / Android 10 Simulator in Test unminified Keymanweb" Web test page using Chrome Browser and I noticed that the bottom of the Key Preview with the bottom of the corresponding key has been detached. 
 



* **TEST_TAMIL_NORMAL_PREVIEW (FAILED):** Tested this in API 29 / Android 10 Simulator in Test unminified Keymanweb" Web test page using Chrome Browser and I noticed that the bottom of the Key Preview with the bottom of the corresponding key has been detached. 
 

* **TEST_CHEROKEE_TOP_PREVIEW (FAILED):** Tested this in API 29 / Android 10 Simulator in Test unminified Keymanweb" Web test page using Chrome Browser and I noticed that the bottom of the Key Preview with the bottom of the corresponding key has been detached. 
 


* **TEST_CHEROKEE_NORMAL_PREVIEW (FAILED):** Tested this in API 29 / Android 10 Simulator in Test unminified Keymanweb" Web test page using Chrome Browser and I noticed that the bottom of the Key Preview with the bottom of the corresponding key has been detached.

**Comment by jahorton** [comment]
Huh... those results without predictive text on Android are certainly odd... I didn't even change any code that should've affected that.

Might be worth verifying that with a check against `stable-15.0`.

**Comment by jahorton** [comment]
:facepalm:  Silly me forgot to change up the group names for in-app vs browser tests.

Guess I'll patch that up manually...

**Comment by jahorton** [comment]
So... when the keytip got done, apparently the 'cap' - the popup-part below - didn't actually get any styling in kmwosk.css, unlike the iOS version.  Without it, we get an invisible one.  Simple enough to fix; patch coming.

**Comment by jahorton** [comment]
@keymanapp-test-bot retest SUITE_1 GROUP_ANDROID_BROWSER all

**Comment by jahorton** [comment]
OK... might have pulled the trigger a little too quickly on the Android key-tip bit.  It definitely needs some extra styling work, so I'll need to figure out how to fix that up.  Root cause was right; it's just that there's extra work beyond it that needs doing.

image

**Comment by jahorton** [comment]
Okay, all better now.

@keymanapp-test-bot retest SUITE_1 GROUP_ANDROID_BROWSER all

**Comment by bharanidharanj** [comment]
## SUITE_1: no predictive text

### GROUP_ANDROID_BROWSER: - use Chrome on a (simulated or real) Android device - your pick.

* **TEST_TAMIL_TOP_PREVIEW (PASSED):** Tested this in API 29 / Android 10 Simulator in Test unminified Keymanweb" Web test page using Chrome Browser and it is working as expected. 
* **TEST_TAMIL_NORMAL_PREVIEW (PASSED):** Tested this in API 29 / Android 10 Simulator in Test unminified Keymanweb" Web test page using Chrome Browser and it is working as expected.


* **TEST_CHEROKEE_TOP_PREVIEW (PASSED):** Tested this in API 29 / Android 10 Simulator in Test unminified Keymanweb" Web test page using Chrome Browser and it is working as expected.
* **TEST_CHEROKEE_NORMAL_PREVIEW (PASSED):** Tested this in API 29 / Android 10 Simulator in Test unminified Keymanweb" Web test page using Chrome Browser and it is working as expected.

**Comment by jahorton** [comment]
The 🍒 is up at #6795.

**Comment by mcdurdin** [comment]
I'm not 100% sure of the test results. https://github.com/keymanapp/keyman/pull/6784#issuecomment-1157562815 shows "PASSED" with screenshots that look clearly wrong to me, e.g. TEST_TAMIL_NORMAL_PREVIEW which has the following associated screenshot, annotated here:

image

**Comment by jahorton** [comment]
Given @mcdurdin's note above...

SUITE_1

GROUP_IPHONE_SE_2_APP

TEST_TAMIL_NORMAL_PREVIEW: OPEN
TEST_CHEROKEE_NORMAL_PREVIEW: OPEN

They still pass on my local machine via local build, but the attached screenshots, as noted, should not have been marked as PASSing.  So, we should double-check that they work properly for our testers, too.  (It shouldn't be related, but note that my local tests are targeting iOS 14.4, rather than iOS 15.x.)

image

image

**Comment by bharanidharanj** [comment]
## SUITE_1: no predictive text

### GROUP_IPHONE_SE_2_APP: - use the Keyman app with a (simulated or real) iPhone SE 2nd gen.

* **TEST_TAMIL_NORMAL_PREVIEW (PASSED):** Retested this in iOS 14.4 / iPhone SE 2nd gen Simulator and it is working as expected. 
 


* **TEST_CHEROKEE_NORMAL_PREVIEW (PASSED):** Retested this in iOS 14.4 / iPhone SE 2nd gen Simulator and it is working as expected.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 16.0.16-alpha](https://keyman.com/downloads/releases/alpha/16.0.16)

**Pull request merged**